File: Jewels_and_Stones/Solution1.java
/* 
	Solution 1 
	
	Use a boolean array to map characters (jewels) that are valid. Check the stone characters against
	the boolean array to find if they are valid and incremement a counter if they are.

	Note: the boolean array is assuming ASCII characters (128)
*/


/* 
	Time Complexity: O(j + s), where j and s are the lengths of the strings jewels and stones
	Space Complexity: O(1)
*/

public class Solution {
    public int NumJewelsInStones(string jewels, string stones) {
        // the number of jewels in your stones bag
        int count = 0;
        
        // Use a boolean array for the characters that are jewels
        bool[] j = new bool[128];
        
        // Insert the jewels into the boolean array
        for (int i = 0; i < jewels.Length; i++){
            j[jewels[i]] = true;
        }
        
        // Loop through the stones and increment if they are in the bool array
        for (int i = 0; i < stones.Length; i++){
            if(j[stones[i]]) count++;
        }
        
        // Return the count of jewels
        return count;
    }
}

/*
	LEET CODE RESULTS

	Runtime: 126 ms, faster than 6.92% of C# online submissions for Jewels and Stones.
	Memory Usage: 23.4 MB, less than 99.79% of C# online submissions for Jewels and Stones.
*/

File: Smaller_Numbers/Solution1.cs
/* 
	Solution 1 
	
	1) Create a frequency table the size of valid numbers (0 through 101).
	2) For each number in nums, increment freqTable[i] by one.
	3) Loop through the freq table, if freqTable[i] > 0, store the value at freqTable[i] in temp. 
		Then, set freqTable[i] to prev and set prev = freqTable[i] + temp.
	4) For each number in nums, nums[i] = freqTable[nums[i]]
*/


/* 
	Time Complexity: O(n) time
	Space Complexity: O(1), since we have a constant size for the freq table
*/

public class Solution {
    public int[] SmallerNumbersThanCurrent(int[] nums) {
        // Frequency Table for valid nums[i]
        int[] freqTable = new int[101];
        int prev = 0;
        int temp = 0;
        
        for (int i = 0; i < nums.Length; i++) {
            freqTable[nums[i]]++;
        }
        
        for (int i = 0; i < 101; i++) {
            if(freqTable[i] != 0){
                temp = freqTable[i];
                freqTable[i] = prev;
                prev = freqTable[i] + temp;
            }
        }
    
        for (int i = 0; i < nums.Length; i++){
            nums[i] = freqTable[nums[i]];
        }
    
        return nums;
        
    }
}

/*
	LEET CODE RESULTS

	Runtime: 236 ms, faster than 95.14% of C# online submissions for How Many Numbers Are Smaller 
		Than the Current Number.
	Memory Usage: 32 MB, less than 97.71% of C# online submissions for How Many Numbers Are Smaller 
		Than the Current Number.
*/